File: 1568/solution.ts
function minDays(grid: number[][]): number {
  const ROWS = grid.length;
  const COLS = grid[0].length;
  const directions = [
    [1, 0],
    [0, 1],
    [-1, 0],
    [0, -1],
  ];
  let islands = 0;
  const visited = new Set<string>();

  function dfs(row: number, col: number, visited: Set<string>): void {
    if (
      row < 0 ||
      row >= ROWS ||
      col < 0 ||
      col >= COLS ||
      visited.has(`${row},${col}`) ||
      grid[row][col] === 0
    ) {
      return;
    }
    visited.add(`${row},${col}`);
    for (const [x, y] of directions) {
      dfs(row + x, col + y, visited);
    }
  }

  // Count the number of islands
  for (let row = 0; row < ROWS; row++) {
    for (let col = 0; col < COLS; col++) {
      if (grid[row][col] === 1 && !visited.has(`${row},${col}`)) {
        islands++;
        dfs(row, col, visited);
      }
    }
  }

  if (islands !== 1) {
    return 0;
  }
  if (visited.size <= 2) {
    return visited.size;
  }

  let minConnections = Infinity;

  // Calculate the minimum connections for any land cell
  for (let row = 0; row < ROWS; row++) {
    for (let col = 0; col < COLS; col++) {
      if (visited.has(`${row},${col}`)) {
        let connections = 0;
        for (const [x, y] of directions) {
          const newRow = row + x;
          const newCol = col + y;
          if (
            newRow >= 0 &&
            newRow < ROWS &&
            newCol >= 0 &&
            newCol < COLS &&
            grid[newRow][newCol] === 1
          ) {
            connections++;
          }
        }
        minConnections = Math.min(minConnections, connections);
      }
    }
  }

  if (minConnections === 1) {
    return 1;
  }

  // Check if removing one land cell disconnects the island
  for (const cell of visited) {
    const [x, y] = cell.split(",").map(Number);
    for (const [r, c] of directions) {
      const nr = x + r;
      const nc = y + c;
      if (nr < 0 || nr >= ROWS || nc < 0 || nc >= COLS || grid[nr][nc] === 0) {
        continue;
      } else {
        grid[x][y] = 0;
        const newVisited = new Set<string>();
        dfs(nr, nc, newVisited);
        if (newVisited.size !== visited.size - 1) {
          return 1;
        }
        grid[x][y] = 1;
      }
    }
  }

  return 2;
}

File: 1568/solution.py
class Solution:
    def minDays(self, grid: list[list[int]]) -> int:
        ROWS, COLS = len(grid), len(grid[0])
        directions = [[1, 0], [0, 1], [-1, 0], [0, -1]]
        islands = 0
        visited = set()

        def dfs(row, col, visited):
            if (
                row < 0
                or row == ROWS
                or col < 0
                or col == COLS
                or (row, col) in visited
                or grid[row][col] == 0
            ):
                return
            visited.add((row, col))
            for x, y in directions:
                dfs(row + x, col + y, visited)

        # Count the number of islands
        for row in range(ROWS):
            for col in range(COLS):
                if grid[row][col] == 1 and (row, col) not in visited:
                    islands += 1
                    dfs(row, col, visited)

        if islands != 1:
            return 0
        if len(visited) <= 2:
            return len(visited)

        min_connections = float("inf")
        land_cells = 0

        # Calculate the minimum connections for any land cell
        for row in range(ROWS):
            for col in range(COLS):
                if (row, col) in visited:
                    land_cells += 1
                    connections = 0
                    for x, y in directions:
                        newRow, newCol = row + x, col + y
                        if (
                            0 <= newRow < ROWS
                            and 0 <= newCol < COLS
                            and grid[newRow][newCol] == 1
                        ):
                            connections += 1
                    min_connections = min(min_connections, connections)

        if min_connections == 1:
            return 1

        # Check if removing one land cell disconnects the island
        for x, y in visited:
            for r, c in directions:
                nr, nc = x + r, y + c
                if nr < 0 or nr == ROWS or nc < 0 or nc == COLS or grid[nr][nc] == 0:
                    continue
                else:
                    grid[x][y] = 0
                    new_visited = set()
                    dfs(nr, nc, new_visited)
                    if len(new_visited) != len(visited) - 1:
                        return 1
                    grid[x][y] = 1

        return 2
